internal/search/parser.go

```diff
@@ -202,15 +207,80 @@ func (p *parser) ParseParameter() Parameter {
 	return ScanParameter(p.buf[start:p.pos])
 }
 
+func visit(node Node, f func(node Node)) {
+	switch v := node.(type) {
+	case Parameter:
+		f(v)
+	case Operator:
+		f(v)
+		for _, n := range v.Operands {
+			visit(n, f)
+		}
+	}
+}
+
+// containsPattern returns true if any descendent of node is a search pattern
+// (i.e., a parameter where the field is the empty string).
+func containsPattern(node Node) bool {
+	var result bool
+	f := func(node Node) {
+		switch v := node.(type) {
+		case Parameter:
+			if v.Field == "" {
+				result = true
+			}
+		}
+	}
+	visit(node, f)
+	return result
+}
+
+// partitionParameters constructs a parse tree to distinguish terms where
+// ordering is insignificant (e.g., "repo:foo file:bar") versus terms where
+// ordering may be significant (e.g., search patterns like "foo bar"). Search
+// patterns are parameters whose field is the empty string.
+//
+// The resulting tree defines an ordering relation on nodes in the following cases:
+// (1) When more than one search patterns exist at the same operator level, they
+// are concatenated in order.
+// (2) Any nonterminal node is concatenated (ordered in the tree) if its
+// descendents contain one or more search patterns.
+func partitionParameters(nodes []Node) []Node {
+	patterns := []Node{}
```

Comment:
https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

```suggestion
	var patterns, unorderedParams []Node
```